File: src/main/java/no/nav/syfo/service/OversikthendelseService.java
package no.nav.syfo.service;

import no.nav.syfo.domain.rest.NyttMotebehov;
import no.nav.syfo.kafka.producer.OversikthendelseProducer;
import no.nav.syfo.kafka.producer.model.KOversikthendelse;
import org.springframework.stereotype.Service;

import java.util.function.Function;

import static java.time.LocalDateTime.now;
import static no.nav.syfo.kafka.producer.OversikthendelseType.MOTEBEHOV_SVAR_BEHANDLET;
import static no.nav.syfo.kafka.producer.OversikthendelseType.MOTEBEHOV_SVAR_MOTTATT;
import static no.nav.syfo.util.MapUtil.map;

@Service
public class OversikthendelseService {

    private OversikthendelseProducer oversikthendelseProducer;

    public OversikthendelseService(OversikthendelseProducer oversikthendelseProducer) {
        this.oversikthendelseProducer = oversikthendelseProducer;
    }

    public void sendOversikthendelse(NyttMotebehov nyttMotebehov) {
        KOversikthendelse kOversikthendelse = map(nyttMotebehov, nyttMotebehov2KOversikthendelse);
        oversikthendelseProducer.sendOversikthendelse(kOversikthendelse);
    }

    public void sendOversikthendelse(String fnr, String tildeltEnhet) {
        KOversikthendelse kOversikthendelse = KOversikthendelse.builder()
                .fnr(fnr)
                .hendelseId(MOTEBEHOV_SVAR_BEHANDLET.name())
                .enhetId(tildeltEnhet)
                .tidspunkt(now()
                ).build();
        oversikthendelseProducer.sendOversikthendelse(kOversikthendelse);
    }

    public static Function<NyttMotebehov, KOversikthendelse> nyttMotebehov2KOversikthendelse = nyttMotebehov -> KOversikthendelse.builder()
            .fnr(nyttMotebehov.arbeidstakerFnr)
            .hendelseId(MOTEBEHOV_SVAR_MOTTATT.name())
            .enhetId(nyttMotebehov.tildeltEnhet)
            .tidspunkt(now())
            .build();
}

File: src/main/kotlin/no/nav/syfo/oversikthendelse/OversikthendelseService.kt
package no.nav.syfo.oversikthendelse

import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.util.*
import javax.inject.Inject

@Service
class OversikthendelseService @Inject constructor(
    private val oversikthendelseProducer: OversikthendelseProducer
) {
    fun sendOversikthendelseMottatt(
        motebehovUUID: UUID,
        arbeidstakerFnr: String,
        tildeltEnhet: String
    ) {
        sendOversikthendelse(motebehovUUID, arbeidstakerFnr, tildeltEnhet, OversikthendelseType.MOTEBEHOV_SVAR_MOTTATT)
    }

    fun sendOversikthendelseBehandlet(
        motebehovUUID: UUID,
        arbeidstakerFnr: String,
        tildeltEnhet: String
    ) {
        sendOversikthendelse(motebehovUUID, arbeidstakerFnr, tildeltEnhet, OversikthendelseType.MOTEBEHOV_SVAR_BEHANDLET)
    }

    private fun sendOversikthendelse(
        motebehovUUID: UUID,
        fnr: String,
        tildeltEnhet: String,
        oversikthendelseType: OversikthendelseType
    ) {
        val kOversikthendelse = map2KOversikthendelse(fnr, tildeltEnhet, oversikthendelseType)
        oversikthendelseProducer.sendOversikthendelse(motebehovUUID, kOversikthendelse)
    }

    private fun map2KOversikthendelse(
        fnr: String,
        tildeltEnhet: String,
        oversikthendelseType: OversikthendelseType
    ) = KOversikthendelse(
        fnr = fnr,
        hendelseId = oversikthendelseType.name,
        enhetId = tildeltEnhet,
        tidspunkt = LocalDateTime.now()
    )
}
